hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,745 @@
+    List<Path> paths =
+        Lists.newArrayList(pathHelper.getTmpNsDir(namespace), pathHelper.getDataNsDir(namespace),
+          pathHelper.getMobDataNsDir(namespace), pathHelper.getArchiveNsDir(namespace));
+    if (includeNamespaceSnapshot) {
+      paths.addAll(getNamespaceSnapshotPath(namespace));
+    }
+    return paths;
+  }
+
+  private List<Path> getNamespaceSnapshotPath(String namespace) throws IOException {
+    return getNamespaceSnapshots(namespace).stream().map(snap -> pathHelper.getSnapshotDir(snap))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * return paths that user will table permission will visit
+   * @param tableName the table
+   * @param includeTableSnapshotPath true if return the paths of table snapshots
+   * @return the path list
+   * @throws IOException if an error occurred
+   */
+  private List<Path> getDefaultTablePath(TableName tableName, boolean includeTableSnapshotPath)
+      throws IOException {
+    List<Path> paths = Lists.newArrayList(pathHelper.getTmpTableDir(tableName),
+      pathHelper.getDataTableDir(tableName), pathHelper.getMobTableDir(tableName),
+      pathHelper.getArchiveTableDir(tableName));
+    if (includeTableSnapshotPath) {
+      paths.addAll(getTableSnapshotPath(tableName));
+    }
+    return paths;
+  }
+
+  private List<Path> getTableSnapshotPath(TableName tableName) throws IOException {
+    return getTableSnapshots(tableName).stream().map(snap -> pathHelper.getSnapshotDir(snap))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Return users with namespace read permission
+   * @param namespace the namespace
+   * @return users with namespace read permission
+   * @throws IOException if an error occurred
+   */
+  private Set<String> getUsersWithNamespaceReadAction(String namespace) throws IOException {
+    return PermissionStorage.getNamespacePermissions(conf, namespace).entries().stream()
+        .filter(entry -> entry.getValue().getPermission().implies(READ))
+        .map(entry -> entry.getKey()).collect(Collectors.toSet());
+  }
+
+  /**
+   * Return users with table read permission
+   * @param tableName the table
+   * @return users with table read permission
+   * @throws IOException if an error occurred
+   */
+  private Set<String> getUsersWithTableReadAction(TableName tableName) throws IOException {
+    return PermissionStorage.getTablePermissions(conf, tableName).entries().stream()
+        .filter(entry -> entry.getValue().getPermission().implies(READ))
+        .map(entry -> entry.getKey()).collect(Collectors.toSet());
+  }
+
+  private List<String> getNamespaceSnapshots(String namespace) throws IOException {
+    return getSnapshots((snapDesc) -> TableName.valueOf(snapDesc.getTable()).getNamespaceAsString()
+        .equals(namespace));
+  }
+
+  private List<String> getTableSnapshots(TableName tableName) throws IOException {
+    return getSnapshots((snapDesc) -> TableName.valueOf(snapDesc.getTable()) == tableName);
```

Comment:
```
(snapDesc) -> TableName.valueOf(snapDesc.getTable()) == tableName
```
Use equals here ?   == only check the hashCode .....

Reply:
done